realkd/estimators.py

```diff
@@ -0,0 +1,357 @@
+    >>> titanic = pd.read_csv('../datasets/titanic/train.csv')
+    >>> survived = titanic.Survived
+    >>> titanic.drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin', 'Survived'], inplace=True)
+    >>> re = RuleBoostingEstimator(base_learner=XGBRuleEstimator(loss=logistic_loss))
+    >>> re.fit(titanic, survived.replace(0, -1), verbose=0) # doctest: +SKIP
+       -1.4248 if Pclass>=2 & Sex==male
+       +1.7471 if Pclass<=2 & Sex==female
+       +2.5598 if Age<=19.0 & Fare>=7.8542 & Parch>=1.0 & Sex==male & SibSp<=1.0
+
+    Multiple base learners can be specified and are used sequentially. The last based learner is used as many times
+    as necessary to learn the desired number of rules. This mechanism can, e.g., be used to fit an "offset rule":
+
+    >>> re_with_offset = RuleBoostingEstimator(num_rules=2, base_learner=[XGBRuleEstimator(loss='logistic', query = Conjunction([])), XGBRuleEstimator(loss='logistic')])
+    >>> re_with_offset.fit(titanic, survived.replace(0, -1)).rules_
+       -0.4626 if True
+       +2.3076 if Pclass<=2 & Sex==female
+
+    >>> greedy = RuleBoostingEstimator(num_rules=3, base_learner=XGBRuleEstimator(loss='logistic', search='greedy'))
+    >>> greedy.fit(titanic, survived.replace(0, -1)).rules_ # doctest: -SKIP
+       -1.4248 if Pclass>=2 & Sex==male
+       +1.7471 if Pclass<=2 & Sex==female
+       -0.4225 if Parch<=1.0 & Sex==male
+    >>> roc_auc_score(survived, greedy.rules_(titanic))
+    0.8321136782454011
+    >>> opt = RuleBoostingEstimator(num_rules=3, base_learner=XGBRuleEstimator(loss='logistic', search='exhaustive'))
+    >>> opt.fit(titanic, survived.replace(0, -1)).rules_ # doctest: -SKIP
+       -1.4248 if Pclass>=2 & Sex==male
+       +1.7471 if Pclass<=2 & Sex==female
+       +2.5598 if Age<=19.0 & Fare>=7.8542 & Parch>=1.0 & Sex==male & SibSp<=1.0
+    >>> roc_auc_score(survived, opt.rules_(titanic)) # doctest: -SKIP
+    0.8490530363553084
+    """
+
+    def __init__(self, num_rules=3, base_learner='XGBRuleEstimator', base_learner_params=None,
+                 verbose=False, weight_update_method='no_update', weight_update_method_params=None):
+        """
+
+        :param int num_rules: the desired number of ensemble members
+        :param Estimator|Sequence[Estimator] base_learner: the base learner(s) to be used in each iteration (last base
+                                    learner is used as many times as necessary to fit desired number of rules)
+        :param bool|int verbose: Level of verbosity, theoretically "number of levels deep of printing"
+        :weight_update_method: the method to do the fully-correction
+        :correction_obj_fn: the method to do the fully-correction
+        """
+        if base_learner_params == None:
+            base_learner_params = {'loss':'squared', 'reg':1.0, 'search':'greedy'}
+
+        self.num_rules = num_rules
+        self.base_learner = SINGLE_RULE_ESTIMATORS[base_learner](**base_learner_params)
```

Comment:
NOTE: this change means that RuleBoostingEstimator can now only take a single learner as input, rather than a collection. Fan concluded that this feature was unused and could be removed, but just wanted to double check (it does seem like a slightly strange feature).

Reply:
NOTE: I decided to revert these changes due to Mario's wise words about keeping this merge on topic. This can be changed in a future PR if needed.

Reply:
This is an important features to allow the combination of different base learners. In particular, it allows enforcing the inclusion of a 'default rule'. @fyan102 , please be careful when deciding which features are unused. Generally, we do not want to give up on any feature without discussion.